### PR TITLE
CFIN-391 Add performance checking to workflow

### PR DESCRIPTION
**Pull request by ChristianBeddows**
Please see the status checks for this pull request :)

One of the concerns about this task is that the result of a lighthouse report can vary between runs, with the same codebase. As such, I've done some investigation using both this script and the browser tool to see if variance is an issue here.

Test figures

From these results I note:
- The scores my script is providing are adequately similar to the scores given by the browser devtool
- The scores from my script have a low variance (seeming lower than the browser tool)

I would therefore conclude that this workflow works and provides an accurate and acceptable score - as opposed to the manual scripts, which still need some work.

### CODE REVIEW

**Review by ncn1 — APPROVED** [review]
Impressive!

**Review by ce666-york — APPROVED** [review]
Great! I've made a couple of suggestions that I think are worth doing, happy for this to be merged though :1st_place_medal: :rocket:

File: src/tests/performance/lighthouseTesting.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+const chromeLauncher = require("chrome-launcher");
+const lighthouse = require("lighthouse");
+const { computeMedianRun } = require("lighthouse/lighthouse-core/lib/median-run");
+const mobileConfig = require("lighthouse/lighthouse-core/config/lr-mobile-config");
+const desktopConfig = require("lighthouse/lighthouse-core/config/lr-desktop-config");
+
+const TEST_URL = "http://localhost:3000?search=maths";
+const ITERATIONS = 5;
+const MINIMUM_DESKTOP_SCORE = 90;
+const MINIMUM_MOBILE_SCORE = 80;
```
[the remark this answers is not in the document]
Comment (ce666-york) [reply src/tests/performance/lighthouseTesting.js:11]: Though I'm wondering how `actions.info` and `actions.setFailed` work locally

File: .gitignore [hunk 2]
```diff
@@ -39,3 +39,6 @@ yarn-error.log*
 
 # serverless
 .serverless
+
+# lighthouse reporting
+lighthouse.*.html
```
Comment (ce666-york) [inline .gitignore:44]: We should exclude these from the package in `serverless.yml` too

File: package.json [hunk 3]
```diff
@@ -17,7 +18,13 @@
     "check": "npm run checkformat && npm run lint && npm run test",
     "formatandcheck": "npm run format && npm run lint && npm run test",
     "fc": "npm run formatandcheck",
-    "percy": "percy exec -- node src/tests/visual/percyVisualTesting.js"
+    "percy": "percy exec -- node src/tests/visual/percyVisualTesting.js",
+    "lighthouse:desktop": "lighthouse --output-path=./lighthouse.desktop.html --emulated-form-factor=desktop --chrome-flags=\"--headless\" http://localhost:3000/?search=maths --view",
+    "lighthouse:mobile": "lighthouse --output-path=./lighthouse.mobile.html --emulated-form-factor=mobile --chrome-flags=\"--headless\" http://localhost:3000/?search=maths --view",
```
Comment (FergusMcGlynn) [inline package.json:23]: These lines come from a separate PR - which makes me wonder whether this PR (or the other one) is targeted at the correct base branch?

**Comment by ChristianBeddows** [comment]
This PR has ultimately passed but I'm re-writing git history a little so I'm going to close this rather than merge.

**Comment by FergusMcGlynn** [comment]
> This PR has ultimately passed but I'm re-writing git history a little so I'm going to close this rather than merge.

FWIW I've found that rewriting history and force pushing it to a branch that's being PR'd works fine. I guess depending on how much rewriting you do, and whether it changes the nature of the PR, it might be worth creating a fresh PR.

File: src/tests/performance/lighthouseTesting.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+(async () => {
+    const isMobileReport = process.argv.includes("--mobile");
+    const chrome = await chromeLauncher.launch({ chromeFlags: ["--headless"] });
```
[the remark this answers is not in the document]
Comment (ChristianBeddows) [reply src/tests/performance/lighthouseTesting.js:15]: Chris has confirmed that a Chrome installation is a pre-requisite

[the remark this answers is not in the document]
Comment (ce666-york) [reply src/tests/performance/lighthouseTesting.js:15]: Answer: it does!

**Pull request closed without merging**